diagnostics-app/src/app/diagnostics/diagnostics-card/diagnostics-card.component.ts

```diff
@@ -0,0 +1,178 @@
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+/**
+ * @fileoverview Defines the Card component for diagnostics.
+ * Imported by diagnostics.module.ts
+ */
+
+import { Component, Input, OnInit, OnDestroy } from '@angular/core';
+import { KeyValue } from '@angular/common';
+
+import {
+  DiagnosticsCardState,
+  DiagnosticsParams,
+  GetRoutineUpdateResponse,
+  RESUMABLE_ROUTINES,
+  RoutineStatus,
+  RoutineType,
+} from '@common/dpsl'
+import {
+  ResponseErrorInfoMessage,
+  RoutineUpdateResponse,
+} from '@common/message';
+import { DiagnosticsService } from 'src/app/core/services/diagnostics.service';
+import { defaultDiagnosticsRefreshInterval, refreshIntervals } from '../../core/config/data-refresh-intervals';
+import { diagnosticsParams } from '../../core/config/diagnostics-params';
+
+@Component({
+  selector: 'app-diagnostics-card',
+  templateUrl: './diagnostics-card.component.html',
+  styleUrls: ['./diagnostics-card.component.css']
+})
+export class DiagnosticsCardComponent implements OnInit, OnDestroy {
+  @Input({ required: true }) routine!: RoutineType;
+
+  private _error?: String; // the error message received, undefined if no error occurs
+  private _intervalId?: number; // the interval id, undefined if there is no interval id
+  private _params?: DiagnosticsParams; // the parameter for starting a routine
+  private _routineId?: number; // the routine id of the running routine
+  private _routineInfo?: GetRoutineUpdateResponse; // the information about the routine
+
+  private _state: DiagnosticsCardState = DiagnosticsCardState.READY; // the state of the card component
+  private _reachedTerminalState = false; // true if the routine reaches a terminal state
+
+  get error() { return this._error; }
+  get intervalId() { return this._intervalId; }
+  get params() { return this._params; }
+  get routineInfo() { return this._routineInfo; }
+  get state() { return this._state; }
+  get messageColor() {
+    switch (this._routineInfo?.status) {
+      case RoutineStatus.passed:
+        return 'green';
+      case RoutineStatus.waiting_user_action:
+        return 'yellow';
+      default:
+        return 'red';
+    }
+  }
+
+  constructor(private diagnosticsService: DiagnosticsService) {}
+
+  ngOnInit(): void {
+    if (diagnosticsParams.has(this.routine)) {
+      this._params = diagnosticsParams.get(this.routine);
+    }
+  }
+
+  ngOnDestroy(): void {
+    if (this._intervalId) {
+      window.clearInterval(this._intervalId);
+    }
+  }
+
+  private handleResponse() {
+    if (!this._routineInfo) {
+      this._error = ResponseErrorInfoMessage.InvalidDiagnosticsRoutineInfo;
+      return;
+    }
+
+    if (this._routineInfo.status === RoutineStatus.waiting_user_action) {
+      this._state = DiagnosticsCardState.WAITING_FOR_USER_ACTION;
+    } else if (
+      this._routineInfo.status === RoutineStatus.passed          ||
```

Comment:
How about use a set and check `if (set.has()) ...```

Reply:
Sure! thanks for the advice.